# Lesson 2

## Loading libraries

In [1]:
import pandas as pd
import numpy as np
import datetime
import warnings
warnings.filterwarnings('ignore')
warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Loading data

In [6]:
data = pd.read_csv('/content/drive/MyDrive/UNIT4/DAY2/Afternoon/unit4_after_morning.csv')
data.columns
data.drop(columns='Unnamed: 0',inplace = True)
data.head()

NameError: ignored

## $\chi^{2}$ test in Python

Let's use the $\chi^{2}$ test in order to determine of columns `DOMAIN` and `RFA_2` are related.

In [9]:
data_crosstab = pd.crosstab(data['DOMAIN'], data['RFA_2'], margins = False)
data_crosstab

RFA_2      L1E   L1F   L1G   L2E   L2F   L2G  ...  L3F  L3G   L4D  L4E  L4F  L4G
DOMAIN                                        ...                               
City       911  5580  2413   962  2063   868  ...  640  303   906  732  358  164
Rural     1092  5617  2035  1034  2010   801  ...  639  253  1042  739  378  146
Suburban   828  6533  2917   828  2293  1010  ...  795  323   811  749  404  189
Town       910  5652  2189   927  2018   866  ...  657  245   912  726  399  155
Urban      489  3736  1529   526  1258   524  ...  367  180   399  435  250  123

[5 rows x 14 columns]

Based on the data in the contingency table we calculate the expected value of the nominal variables. Based on the expected values, the chi-square test statistic is calculated which helps us decide on whether the variables are independent or not. Technically, though the value of the test statistic we are trying to prove or disprove some hypotheses on the independence of categorical variables.

* **H0** (Null Hypothesis) - assumes that there is no association between the two variables.

* **Ha** (Alternate Hypothesis) - assumes that there is an association between the two variables.

In [ ]:
from scipy.stats import chi2_contingency

chi2, p_value, dof, expected_freq = chi2_contingency(data_crosstab, correction=False)

print("The Chi2 value is: ",round(chi2,2))
print("The p-value is: ",p_value)
print("The number of degrees of freedom is: ",dof)
print("The expected frequencies are: ")
pd.DataFrame(expected_freq, columns=data_crosstab.columns, index=data_crosstab.index)

The Chi2 value is:  464.56
The p-value is:  1.349917635349138e-67
The number of degrees of freedom is:  52
The expected frequencies are: 


RFA_2            L1E          L1F  ...         L4F         L4G
DOMAIN                             ...                        
City      897.919336  5756.448360  ...  379.758320  164.936956
Rural     899.840777  5768.766472  ...  380.570957  165.289902
Suburban  991.260878  6354.849288  ...  419.235393  182.082672
Town      891.699938  5716.576575  ...  377.127941  163.794528
Urban     549.279071  3521.359305  ...  232.307390  100.895943

[5 rows x 14 columns]

We can determine if we accept the the null hypotheses by two ways:

* Comparing the Chi2 againts the Chi2_critical value
* Comparing the p-value againt the corresponding p_critical

Let's use the usual confidence level of 95%.

* Using the Chi2 table:

We look for dof = 52 and p = 0.05 give us: 69.83.
Now we compare our Chi2 against the critical value of chi2
which is: 69.83

464.56 > 69.83  -> We **reject** the nul hypotheses **Ho**

* Using the p_value:

We compare the our p_value against 0.05 (95%)

0 < 0.05 -> -> We **reject** the nul hypotheses **Ho**

Remember the meaning of the p-value = P(D|Ho): the probability of observing our data under the null hypotheses.

Therefore, **columns `RFA_2` and `DOMAIN` are related.**

## Activity 

Keep working with unit4.csv that you already have locally.

Use the Chi-Square test for measuring the salary differences (ie. `INCOME`) between men and women (assume a confidence level of 95%)

First group income in two groups:

* INCOME > 3 -> 'HIGH_INCOME'
* INCOME <= 3 -> 'LOW_INCOME'

* If the observed chi-square test statistic is greater than the critical value (this value is known already based on certain parameters) in the data, the null hypothesis can be rejected.

* If the observed chi-square test statistic is lower than the critical value (this value is known already based on certain parameters) in the data, the null hypothesis is accepted (also put as we fail to reject the null hypothesis) 

Chi2 <-> Chi2_critical

```python
if (Chi2 >= Chi2_critical):
    reject Ho
else:
    accept H0
```

Based on the statistics we either reject H0 or we fail to reject H0. You can also use the p value directly as we will see later in the lesson.



We have seen that in our case:

**columns `RFA_2` and `DOMAIN` are related.**

so let's drop one of those columns, for example RFA_2.

In [ ]:
data.drop('RFA_2', axis = 1, inplace = True)

## Activity

Repeat the same steps already done on the class for the columns `STATE` and `DOMAIN` to find the chi-square statistic.

Check the p value to decide to reject or fail to reject the null hypothesis. If you reject the null hypothesis, then drop one of the columns (`STATE` here).

You can get the critical values for the Chi-square distribution with n degrees of freedom here (check here)[https://people.smp.uq.edu.au/YoniNazarathy/stat_models_B_course_spring_07/distributions/chisqtab.pdf]:

You can compute evrything with Python using (here)[https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html]

```python
if (p_value < 0.05):
    reject Ho
else:
    accept Ho
``` 

Are columns `STATE` and `DOMNAIN` are related according to the Chi_square test?
